src/app/registration/registration.component.spec.ts

```diff
@@ -0,0 +1,140 @@
+// TODO at a later date if there is time, unit test to make sure that component.formError[field] has a value, not sure how to detect
+// change yet
+
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+import {
+    FormGroup,
+    ReactiveFormsModule
+} from '@angular/forms';
+import { HttpModule } from '@angular/http';
+import { RouterTestingModule } from "@angular/router/testing";
+
+
+import { RegistrationComponent } from './registration.component';
+
+import { RegistrationDirective } from './registration.directive';
+
+import { UserService } from '../shared/services';
+import { ApiService } from '../shared/services';
+import { Observable } from "rxjs/Observable";
+
+describe('RegistrationComponent', () => {
+  let component: RegistrationComponent;
+  let fixture: ComponentFixture<RegistrationComponent>;
+  let userServiceStub: object;
+  let userService: any;
+  let testResponse: any;
+  let spy: any;
+
+
+  beforeEach(async(() => {
+
+    TestBed.configureTestingModule({
+      imports: [ReactiveFormsModule, HttpModule, RouterTestingModule],
+      declarations: [ RegistrationComponent, RegistrationDirective ],
+      providers: [ ApiService, UserService ]
+    }).compileComponents();
+ }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(RegistrationComponent);
+    component = fixture.componentInstance;
+    userService = fixture.debugElement.injector.get(UserService);
+    testResponse = {errors: "", success: false};
+    spy = spyOn(userService, 'createUser').and.returnValue(Observable.of(testResponse));
+    fixture.detectChanges();
+  });
+
+  it('should create the registration component', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('should create a FormGroup', () => {
+    component.ngOnInit();
+    expect(component.user instanceof FormGroup).toBe(true);
+  });
+
+  it('should reject username that are less than 2 characters long', () => {
```

Comment:
We should add a test that tests if the control is `invalid` and/or `dirty` to avoid the issue I discuss in `registration.component.ts`.

Reply:
We can save that for a larger project. I am not expecting 100% code coverage here or anywhere near it.